packages/metadata-editor/src/MetadataEditor.tsx

```diff
@@ -0,0 +1,271 @@
+        this.schema = schema;
+      }
+    }
+
+    this.allMetadata = await FrontendServices.getMetadata(this.namespace);
+    if (this.name) {
+      for (const metadata of this.allMetadata) {
+        if (this.name == metadata.name) {
+          this.metadata = metadata;
+          this.title.label = this.metadata.display_name;
+        }
+      }
+    }
+  }
+
+  onCloseRequest(msg: Message): void {
+    if (this.dirty) {
+      showDialog({
+        title: 'Close without saving?',
+        body: (
+          <p>
+            {' '}
+            {`"${
+              this.getFormItem('Name').value
+            }" has unsaved changes, close without saving?`}{' '}
+          </p>
+        ),
+        buttons: [Dialog.cancelButton(), Dialog.okButton()]
+      }).then((response: any): void => {
+        if (response.button.accept) {
+          this.dispose();
+          super.onCloseRequest(msg);
+        }
+      });
+    } else {
+      this.dispose();
+      super.onCloseRequest(msg);
+    }
+  }
+
+  saveMetadata(): void {
+    const newMetadata: any = {
+      schema_name: this.schemaName,
+      display_name: this.getFormItem('Name').value,
+      metadata: this.metadata
+    };
+
+    for (const field of this.metadataForm) {
+      if (field.type == 'TextInput') {
+        newMetadata.metadata[field.schemaField] = field.value;
+      } else if (field.type == 'DropDown') {
+        newMetadata.metadata[field.schemaField] = field.value.choice;
+      } else if (field.type == 'Code') {
+        newMetadata.metadata[field.schemaField] = field.value.split('\n');
+      }
+    }
+
+    if (!this.name) {
+      FrontendServices.postMetadata(
+        this.namespace,
+        JSON.stringify(newMetadata)
+      ).then((response: any): void => {
+        this.handleDirtyState(false);
+        this.onSave();
+        this.close();
+      });
+    } else {
+      FrontendServices.putMetadata(
+        this.namespace,
+        this.name,
+        JSON.stringify(newMetadata)
+      ).then((response: any): void => {
+        this.handleDirtyState(false);
+        this.onSave();
+        this.close();
+      });
+    }
+  }
+
+  getFormItem(searchLabel: string): FormItem {
+    return this.metadataForm.find(({ value, type, label, schemaField }) => {
+      return label == searchLabel;
+    });
+  }
+
+  handleTextInputChange(event: any, label: string): void {
+    this.handleDirtyState(true);
+    this.getFormItem(label).value = event.nativeEvent.srcElement.value;
+  }
+
+  handleDropdownChange = (label: string, value: string): void => {
+    this.handleDirtyState(true);
+    this.getFormItem(label).value.choice = value;
+    this.update();
+  };
+
+  handleDirtyState(dirty: boolean): void {
+    this.dirty = dirty;
+    if (this.dirty && !this.title.className.includes(DIRTY_CLASS)) {
+      this.title.className += ` ${DIRTY_CLASS}`;
+    } else if (!this.dirty) {
+      this.title.className = this.title.className.replace(DIRTY_CLASS, '');
+    }
```

Comment:
I don't understand the utility for appending/masking `'jp-mod-dirty'` to/from the className of the title.  The className is never checked for this value that I can tell.

What do the back-ticks do on this statement?
```
this.title.className += ` ${DIRTY_CLASS}`;
```

Reply:
We're doing that because it adds an indicator over the "Close tab" button to show that there are unsaved changes to the metadata (this is implemented as part of the `DocumentWidget` code, but as long as it has that class name it can reuse that css for consistency). So before you make any changes to the metadata form, the tab would look like this:
![image](https://user-images.githubusercontent.com/6673460/85622565-a74ed880-b62c-11ea-82df-81198a6b72db.png)
Then after you make changes, it looks like this: 
![image](https://user-images.githubusercontent.com/6673460/85622590-b5045e00-b62c-11ea-9f52-653166f17eae.png)
And if you hover over that black circle it looks like this:
![image](https://user-images.githubusercontent.com/6673460/85622771-f4cb4580-b62c-11ea-90e3-3bae265a528d.png)
And if you press it, it opens a dialog to confirm you want to close without saving changes. 
You're right about the back ticks, I just removed them!

Reply:
Thank you for the great explanation and I am ever more grateful that others deal with this level of detail than myself! :smile: